addons/mass_mailing/models/mailing_trace.py

```diff
@@ -125,13 +126,30 @@ def _compute_display_name(self):
         for trace in self:
             trace.display_name = f'{trace.trace_type}: {trace.mass_mailing_id.name} ({trace.id})'
 
+    @api.depends("res_id")
+    def _compute_recipient(self):
+        for trace in self:
+            recipient = trace.env[trace.model].browse(trace.res_id)
+            trace.recipient = recipient.name
+
     @api.model_create_multi
     def create(self, vals_list):
         for values in vals_list:
             if 'mail_mail_id' in values:
                 values['mail_mail_id_int'] = values['mail_mail_id']
         return super().create(vals_list)
 
+    def action_retry(self):
+        traces = self.filtered(lambda t: t.trace_status in ("error", "cancel", "bounce"))
+        if not traces:
+            return
+        mailings = traces.mapped("mass_mailing_id")
+        mails = traces.mapped("mail_mail_id")
+        mails.unlink()
+        traces.unlink()
+        for mailing in mailings:
+            mailing.action_put_in_queue()
```

Comment:
Can't we use `action_retry_failed` from mailing_mailing ?
like this ?
```suggestion
        mailings = traces.mass_mailing_id.action_retry_failed()
```

Reply:
ok then
```suggestion
        mailings = traces.mass_mailing_id
        traces.mail_mail_id.unlink()
        traces.unlink()
        mailings.action_put_in_queue()
```